### DISCUSSION TITLE
Unable to migrate from webpack 4 to 5.

### DISCUSSION THREAD

**Discussion by Aabir1** [webpack/webpack#13670] [General]
I am very new to webpack. I have to migrate an application which written in Laravel, Docker and below versions.
I read complete documents and tried with stackoverflow as well.
____________________
|   Previous Version        |
---------------------------------
node v9.11.2
npm v5.0.3
babel-loader:  v^6.2.4
gulp v3.9.1
webpack v^3.8.1

preCommand => ./node_modules/.bin/webpack --progress --colors

____________________
|          New Version        |
---------------------------------

node v14.17.1
npm v6.14.13
babel-loader:  v^7.1.4
gulp v4.0.2
webpack v^5.40.0

newCommand => ./node_modules/.bin/webpack --progress --colors

# Bug report

**What is the current behavior?**

resolve 'common/network/Network' in '/web/client/common/redux'
  Parsed request is a module
  using description file: /web/client/package.json (relative path: ./common/redux)
    aliased with mapping 'common': './common' to './common/network/Network'
      using description file: /web/client/package.json (relative path: ./common/redux)
        Field 'browser' doesn't contain a valid alias configuration
        using description file: /web/client/package.json (relative path: ./common/redux/common/network/Network)
          no extension
            Field 'browser' doesn't contain a valid alias configuration
            /web/client/common/redux/common/network/Network doesn't exist
          .js
            Field 'browser' doesn't contain a valid alias configuration
            /web/client/common/redux/common/network/Network.js doesn't exist
          .jsx
            Field 'browser' doesn't contain a valid alias configuration
            /web/client/common/redux/common/network/Network.jsx doesn't exist
          .less
            Field 'browser' doesn't contain a valid alias configuration
            /web/client/common/redux/common/network/Network.less doesn't exist
          .json
            Field 'browser' doesn't contain a valid alias configuration
            /web/client/common/redux/common/network/Network.json doesn't exist
          as directory
            /web/client/common/redux/common/network/Network doesn't exist

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

**Other relevant information:**
webpack version:  
Node.js version: 
Operating System: 
Additional tools:

**Comment by alexander-akait** [comment]
Please rewrite `common/network/Network` on `./common/network/Network.js`

**Reply by Aabir1** [reply]
Hi,

Sorry for delayed reply. I am using `const Network = require("common/network/Network"); ` in most of the internal files and folder.
Do I have to change it in every internal file ?

In webpack.config.js, I am already using `./common/`.

**Reply by alexander-akait** [reply]
`require("common/network/Network")` means load `common` module, you should use `./` to indicate that it is module

**Reply by Aabir1** [reply]
Below is the working code with the previous version

'use strict';

const path = require('path');
const webpack = require('webpack');

const headerPath = './common/pages/header.js';
const connectionLostPath = './common/pages/connectionLost.js';
const acceptNewTermsOfService = './common/pages/acceptNewTermsOfService.js';
const commonComponents = [headerPath, connectionLostPath, acceptNewTermsOfService];

const ExtractTextPlugin = require("extract-text-webpack-plugin");
const LessExtractCss = new ExtractTextPlugin({
	filename:  getPath => getPath('css/[name]-js.css').replace('css', '../../css')
});

function resolve(relPath){
	return path.resolve(__dirname, relPath);
}

const AntdChunks = new webpack.optimize.CommonsChunkPlugin({
	name: 'antd',
	minChunks: module => module.context && module.context.indexOf('antd') !== -1
});

module.exports = {
    context: __dirname,
    entry: {
		// header
		'header' : commonComponents,
		// Auth
        'login': './auth/pages/login.js',
        'register': './auth/pages/register.js',
        'reset-password': './auth/pages/reset-password.js',
		'complete-registration': './auth/pages/complete-registration.js',



        // Common
        'settings': ['./common/pages/settings.js'].concat(commonComponents),

	},
    output: {
        path: resolve('../public/js/pages'),
        filename: "[name].js"
    },
    devtool: 'source-map',
    resolve: {
        modules: ["node_modules", "./"],
        extensions: ['.js', '.jsx', '.less'],
        alias: {
			'react': resolve('node_modules/react'),
			'asn1.js': 'node_modules/asn1.js/'
		}
    },
    plugins: [LessExtractCss],
	resolveLoader: {
		modules: ['node_modules']
	},
    module: {
        rules: [
            {
                test: /.jsx?$/,
                loader: 'babel-loader',
                exclude: /node_modules/,
                query: {
                    presets: ["react", "es2015"]
                }
            }, {
				test: /\.less$/,
				use: ExtractTextPlugin.extract({
					use: [{
						loader: 'css-loader'
					}, {
						loader: "less-loader"
					}, {
						loader: 'antd-loader',
						query: {
							override: {
								srcDir: resolve('./less/antd/'),
								files: [/lib\/style/]
							},
							namespace: {
								selector: '.ns-ant',
								files: [/lib\/form/]
							}
						}
					}]
				})
			}
        ]
    },
    node: {
        fs: "empty"
    }

};

**Reply by alexander-akait** [reply]
`CommonsChunkPlugin` and `ExtractTextPlugin` deprecated you need to migrate on `optimization.splitChunks` and `mini-css-extract-plugin`

**Reply by Aabir1** [reply]
Yes Already migrated, above is my old configurations. See below is the updated configurations

'use strict';

const path = require('path');
const webpack = require('webpack');

const headerPath = './common/pages/header.js';
const connectionLostPath = './common/pages/connectionLost.js';
const acceptNewTermsOfService = './common/pages/acceptNewTermsOfService.js';
const commonComponents = [headerPath, connectionLostPath, acceptNewTermsOfService];

// const ExtractTextPlugin = require("extract-text-webpack-plugin");
// const LessExtractCss = new ExtractTextPlugin({
// 	filename:  getPath => getPath('css/[name]-js.css').replace('css', '../../css')
// });

const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const LessExtractCss = new MiniCssExtractPlugin({
	filename:  getPath => getPath('css/[name]-js.css').replace('css', '../../css')
});


function resolve(relPath){
	return path.resolve(__dirname, relPath);
}

// const AntdChunks = new webpack.optimize.CommonsChunkPlugin({
// 	name: 'antd',
// 	minChunks: module => module.context && module.context.indexOf('antd') !== -1
// });

module.exports = {
    context: __dirname,

    entry: {
		// header
		'header' : commonComponents,

		// Auth
        'login': './auth/pages/login.js',
        'register': './auth/pages/register.js',
        'reset-password': './auth/pages/reset-password.js',
		'change-password': './auth/pages/change-password.js',
		'complete-registration': './auth/pages/complete-registration.js',

    output: {
        path: resolve('../public/js/pages'),
        filename: "[name].js"
    },
    devtool: 'source-map',
    resolve: {
        modules: [
	        path.join(__dirname, 'node_modules')
	    ],
        extensions: ['.js', '.jsx', '.less', '.json'],
        alias: {
			'react': resolve('node_modules/react'),
			'asn1.js': 'node_modules/asn1.js/',

  			'common': './common',
		}
    },
    plugins: [
    	LessExtractCss
    	//  [
	    //   "@babel/plugin-transform-runtime",
	    //   {
	    //     "absoluteRuntime": false,
	    //     "corejs": 3,
	    //     "version": "^7.7.4"
	    //   }
	    // ]
    ],
	resolveLoader: {
		modules: [
        	path.join(__dirname, 'node_modules')
		]
	},
    module: {
        rules: [
            {
                test: /\.jsx?$/, //[/\.js$/, /\.jsx/]
                loader: 'babel-loader',
                exclude: /node_modules/,
                // query: {
                //     presets: ["react", "es2015"]
                // },
                options: {
			      presets: ["react" ,"es2015"]
			    },
            }, {
				test: /\.less$/,
				use: [
					{ loader: 'css-loader' },
					{ loader: "less-loader" },
					{ loader: "style-loader"},
					{
						loader: 'antd-loader',
						options: {
							override: {
								srcDir: resolve('./less/antd/'),
								files: [/lib\/style/]
							},
							namespace: {
								selector: '.ns-ant',
								files: [/lib\/form/]
							}
						}
					}
				]
			},
        ]
    },
    optimization: {
    	splitChunks: {
		  chunks: 'async',
		  minSize: 30000,
		  minRemainingSize: 0,
		  maxSize: 0,
		  minChunks: 1,
		  maxAsyncRequests: 6,
		  maxInitialRequests: 4,
		  automaticNameDelimiter: '~',
		  cacheGroups: {
		    vendors: {
		      test: /[\\/]node_modules[\\/]/,
		    },
		    default: {
		      minChunks: 2,
		      priority: -20,
		      reuseExistingChunk: true
		    }
		  }
		}
    },
 //    resolve: {
	//     fallback: {
	//       fs: false
	//     }
	// }

};

**Reply by alexander-akait** [reply]
Please prefer to use git repo for this, code is unreadable here

**Reply by Aabir1** [reply]
You want me to create a git repo for the above file ?

**Reply by alexander-akait** [reply]
Will be great, it will allow to help you better

**Reply by Aabir1** [reply]
Thanks, Allow me 30minutes, I will create and give it to you.

**Reply by Aabir1** [reply]
@alexander-akait here is the link => https://github.com/Aabir1/testing-webpack

If you want, I can add old configuration as well.

Thanks.

**Reply by alexander-akait** [reply]
Without `package.json` and simple code I can't help, I don't know your versions and other related stuff

**Reply by Aabir1** [reply]
Hi Alex, I have added package.json and some folder structure of my code. Can you please give it a look.

Thanks.

**Reply by alexander-akait** [reply]
I can't install due:
```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@bitbucket.org/personal-code/personal-code-common-components.git
npm ERR! 
npm ERR! git@bitbucket.org: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128
```

Hints:
- update all webpack loaders and plugins to the latest version
- update babel and all babel plugins
- migrate from `optimize-css-assets-webpack-plugin` to https://github.com/webpack-contrib/css-minimizer-webpack-plugin
- you don't need `filename:  getPath => getPath('css/[name]-js.css').replace('css', '../../css')` anymore, publicPath is `auto` by default for webpack v5 and mini-css-extract-plugin supports it, so all assets always relative to CSS files
- don't use ` extensions: ['.js', '.jsx', '.less']`, it is invalid
you have wrong usage of `css-loader`, `less-loader` and `mini-css-extract-plugin`, please look at this https://github.com/webpack-contrib/mini-css-extract-plugin#recommend
- seems `ant-loader` is not supported anymore, repo doesn't exists https://github.com/tenadolanter/antd-loader, so I can't help here... 
- You don't need:
```
config.plugins.push(new webpack.DefinePlugin({
	'process.env': {
		'NODE_ENV': JSON.stringify('production')
	}
}));
```

webpack does it automatically https://github.com/webpack/webpack/blob/main/lib/WebpackOptionsApply.js
- also you have overlapping for `strip-loader` and `babel-loader`, don't use `strip-loader`, you don't need this due `terser-webpack-plugin` https://github.com/webpack-contrib/terser-webpack-plugin/releases have `drop_console` option https://github.com/terser/terser#compress-options, they even have recommendation about this https://github.com/YahooArchive/strip-loader but for `uglify-webpack-plugin` (this plugin also deprecated in favor `terser-webpack-plugin`)
- also all minimizers plugins, i.e. `terser-webpack-plugin`/`optimize-css-assets-webpack-plugin` (I written above about migration) should be in the `minimizer` option, here example how to setup this https://webpack.js.org/configuration/optimization/, so you don't need `webpack-production.config.js` file

**Reply by Aabir1** [reply]
Thanks for this, allow me sometime to play with the changes that you have suggested.
